src/ralph/backends/data/clickhouse.py

```diff
@@ -0,0 +1,524 @@
+
+        If username and password are None, we will try to connect as the ClickHouse
+        user "default".
+        """
+        self.host = settings.HOST
+        self.port = settings.PORT
+        self.database = settings.DATABASE
+        self.event_table_name = settings.EVENT_TABLE_NAME
+        self.username = settings.USERNAME
+        self.password = settings.PASSWORD
+        self.client_options = settings.client_options.dict()
+        self.locale_encoding = settings.LOCALE_ENCODING
+        self._client = None
+
+    @property
+    def client(self):
+        """Creates a ClickHouse client if it doesn't exist.
+
+        We do this here so that we don't interrupt initialization in the case
+        where ClickHouse is not running when Ralph starts up, which will cause
+        Ralph to hang. This client is HTTP, so not actually stateful. Ralph
+        should be able to gracefully deal with ClickHouse outages at all other
+        times.
+        """
+        if not self._client:
+            self._client = clickhouse_connect.get_client(
+                host=self.host,
+                port=self.port,
+                database=self.database,
+                username=self.username,
+                password=self.password,
+                settings=self.client_options,
+            )
+        return self._client
+
+    def status(self) -> DataBackendStatus:
+        """Checks ClickHouse connection status."""
+        try:
+            self.client.query("SELECT 1")
+        except ClickHouseError:
+            return DataBackendStatus.AWAY
+
+        return DataBackendStatus.OK
+
+    def list(
+        self, target: str = None, details: bool = False, new: bool = False
+    ) -> Iterator[Union[str, dict]]:
+        """Lists documents for a given table.
+
+        Args:
+            target (str): The table to list documents from.
+            details (bool): Get detailed archive information instead of just ids.
+            new (bool): Given the history, list only not already fetched archives.
+        """
+        sql = "SHOW TABLES"  # nosec
+        for document in self.client.query(sql).named_results():
+            if details:
+                yield document
+            else:
+                yield str(document.get("name"))
+
+    @enforce_query_checks
+    def read(
+        self,
+        *,
+        query: Union[str, ClickHouseQuery] = None,
+        target: str = None,
+        chunk_size: Union[None, int] = None,
+        raw_output: bool = False,
+        ignore_errors: bool = False,
+    ) -> Iterator[Union[bytes, dict]]:
+        """Reads documents from the database."""
+        reader = self._read_raw if raw_output else self._read_dict
+        fields = ",".join(query.return_fields) if query.return_fields else "event"
+
+        if target is None:
+            target = self.event_table_name
+        sql = f"SELECT {fields} FROM {target}"  # nosec
```

Comment:
Why not use queries with parameters instead?
https://clickhouse.com/docs/en/interfaces/cli#cli-queries-with-parameters